### PR TITLE
.Net: Rename 'Skills' -> 'Plugins' - Part 3

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

Rename Skills to Functions for SK core and abstractions packages. Third sub task from https://github.com/microsoft/semantic-kernel/issues/2119.

### Description

 - Remove the `Microsoft.SemanticKernel.SkillDefinition` namespace and promote classes previously contained to `Microsoft.SemanticKernel`, this is to reduce the number of `using`s that client code required.
 - Rename variable/function names using skillName to pluginName.
 - Rename skill collection interfaces/classes to use function collection.
 - Leave some obsolete methods/properties to help with the upgrade from earlier releases.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/IntegrationTests/Planning/SequentialPlanner/SequentialPlannerTests.cs [hunk 1]
```diff
@@ -39,8 +39,8 @@ public async Task CreatePlanFunctionFlowAsync(bool useChatModel, string prompt,
         // Arrange
         bool useEmbeddings = false;
         IKernel kernel = this.InitializeKernel(useEmbeddings, useChatModel);
-        _ = kernel.ImportSkill(new EmailSkillFake());
-        TestHelpers.GetSkills(kernel, "FunSkill");
+        _ = kernel.ImportPlugin(new EmailSkillFake());
+        TestHelpers.GetPlugins(kernel, "FunSkill");
```
Comment (shawncal) [inline dotnet/src/IntegrationTests/Planning/SequentialPlanner/SequentialPlannerTests.cs:43]: These are weird... why are we calling this but not getting the results? If it's deliberate, can you update with 
`_ = TestHelpers.GetPlugins....`

File: dotnet/src/IntegrationTests/TestHelpers.cs [hunk 2]
```diff
@@ -27,16 +26,16 @@ internal static void ImportSampleSkills(IKernel target)
             "QASkill");
     }
 
-    internal static IDictionary<string, ISKFunction> GetSkills(IKernel target, params string[] skillNames)
+    internal static IDictionary<string, ISKFunction> GetPlugins(IKernel target, params string[] pluginNames)
```
Comment (shawncal) [inline dotnet/src/IntegrationTests/TestHelpers.cs:29]: Ah, can you change to GetOrImportPlugins or something like that, to indicate that it's actually doing the import work too?

File: dotnet/src/SemanticKernel/Kernel.cs [hunk 3]
```diff
@@ -83,67 +88,76 @@ public sealed class Kernel : IKernel, IDisposable
...
-        ISKFunction function = this.CreateSemanticFunction(skillName, functionName, functionConfig);
-        this._skillCollection.AddFunction(function);
+        ISKFunction function = this.CreateSemanticFunction(pluginName, functionName, functionConfig);
+        this._functionCollection.AddFunction(function);
 
         return function;
     }
 
     /// <inheritdoc/>
-    public IDictionary<string, ISKFunction> ImportSkill(object skillInstance, string? skillName = null)
```
Comment (shawncal) [inline dotnet/src/SemanticKernel/Kernel.cs:111]: Some of these are on the critical path for callers... can you restore `ImportSkill` as obsolete and have it call the `ImportPlugin`?

File: dotnet/src/InternalUtilities/test/FunctionHelpers.cs [hunk 4]
```diff
@@ -15,13 +14,13 @@ internal static class FunctionHelpers
     /// Invokes a function on a skill instance via the kernel.
     /// </summary>
     public static Task<SKContext> CallViaKernelAsync(
-        object skillInstance,
+        object objectInstance,
```
Comment (shawncal) [inline dotnet/src/InternalUtilities/test/FunctionHelpers.cs:17]: pluginInstance or pluginObjectInstance?

File: dotnet/src/IntegrationTests/TestHelpers.cs (hunk 2 above)
Comment (markwallace-microsoft) [reply dotnet/src/IntegrationTests/TestHelpers.cs]: Renamed to `ImportSamplePlugins`

File: dotnet/src/SemanticKernel/Kernel.cs (hunk 3 above)
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel/Kernel.cs:111]: Yes, that's already in place. All of the main methods/properties that I changed have an original stub which calls the new method/property.

File: dotnet/src/InternalUtilities/test/FunctionHelpers.cs (hunk 4 above)
Comment (markwallace-microsoft) [reply dotnet/src/InternalUtilities/test/FunctionHelpers.cs]: Used `pluginInstance`

File: dotnet/src/SemanticKernel/Functions/FunctionCollection.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+/// Semantic Kernel default function collection class.
+/// The class holds a list of all the functions, native and semantic, known to the kernel instance.
+/// The list is used by the planner and when executing pipelines of function compositions.
+/// </summary>
+[SuppressMessage("Naming", "CA1711:Identifiers should not have incorrect suffix")]
+[DebuggerTypeProxy(typeof(IReadOnlyFunctionCollectionTypeProxy))]
+[DebuggerDisplay("{DebuggerDisplay,nq}")]
+public class FunctionCollection : IFunctionCollection
+{
+    internal const string GlobalFunctions = "_GLOBAL_FUNCTIONS_";
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel/Functions/FunctionCollection.cs:27]: For legibility

```suggestion
    internal const string GlobalFunctionsCollectionName = "_GLOBAL_FUNCTIONS_";
```

File: dotnet/src/SemanticKernel/Functions/FunctionCollection.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="FunctionCollection"/> class.
+    /// </summary>
+    /// <param name="loggerFactory">The logger factory.</param>
+    public FunctionCollection(ILoggerFactory? loggerFactory = null)
+    {
+        this._logger = loggerFactory is not null ? loggerFactory.CreateLogger(typeof(FunctionCollection)) : NullLogger.Instance;
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel/Functions/FunctionCollection.cs:35]:
```suggestion
        this._logger = loggerFactory?.CreateLogger(typeof(FunctionCollection)) ??  NullLogger.Instance;
```

File: dotnet/src/SemanticKernel/Functions/FunctionCollection.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+    internal string DebuggerDisplay => $"Count = {this._functionCollection.Count}";
+
+    #region private ================================================================================
+
+    [DoesNotReturn]
+    private void ThrowFunctionNotAvailable(string pluginName, string functionName)
+    {
+        this._logger.LogError("Function not available: plugin:{0} function:{1}", pluginName, functionName);
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel/Functions/FunctionCollection.cs:114]: Change to template on affected code.

```suggestion
        this._logger.LogError("Function not available: plugin:{PluginName} function:{FunctionName}", pluginName, functionName);
```

File: dotnet/src/SemanticKernel/Planning/IPlan.cs [hunk 8]
```diff
@@ -1,7 +1,5 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 
-using Microsoft.SemanticKernel.SkillDefinition;
-
 namespace Microsoft.SemanticKernel.Planning;
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel/Planning/IPlan.cs:3]:
```suggestion
using Microsoft.SemanticKernel.Functions;

namespace Microsoft.SemanticKernel.Planning;
```

File: dotnet/src/SemanticKernel/Text/SKFunctionTextExtensions.cs [hunk 9]
```diff
@@ -5,8 +5,8 @@
 using Microsoft.SemanticKernel.Orchestration;
 
 #pragma warning disable IDE0130
-// ReSharper disable once CheckNamespace - Using NS of ISKFunction
-namespace Microsoft.SemanticKernel.SkillDefinition;
+// ReSharper disable once CheckNamespace - Using main namespace
+namespace Microsoft.SemanticKernel;
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel/Text/SKFunctionTextExtensions.cs:9]: Avoid using `Microsoft.SemanticKernel` when the folder don't represent it.

Not sure why is this in a `Text` folder, the correct place should be close to what it is extending (if in the same package)..

```suggestion
namespace Microsoft.SemanticKernel.Functions;
```

**Review by RogerBarreto — CHANGES_REQUESTED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Functions/FunctionView.cs [hunk 10]
```diff
@@ -3,18 +3,21 @@
 using System;
 using System.Collections.Generic;
 
-namespace Microsoft.SemanticKernel.SkillDefinition;
+#pragma warning disable IDE0130
+// ReSharper disable once CheckNamespace - Using the main namespace
+namespace Microsoft.SemanticKernel;
```
[the remark this answers is not in the document]
Comment (markwallace-microsoft) [reply dotnet/src/SemanticKernel.Abstractions/Functions/FunctionView.cs:8]: Shawn asked me to make this change, he wants to reduce the number of namespaces

**Review by RogerBarreto — APPROVED** [review]
LGTM (Except for the Namespaces change)

File: dotnet/src/Extensions/Planning.StepwisePlanner/Planning.StepwisePlanner.csproj [hunk 11]
```diff
@@ -21,8 +21,8 @@
     <InternalsVisibleTo Include="SemanticKernel.Extensions.StepwisePlanner.UnitTests" />
   </ItemGroup>
   <ItemGroup>
-    <None Remove="Skills/**/*" />
-    <EmbeddedResource Include="Skills/**/*">
+    <None Remove="Plugin/**/*" />
+    <EmbeddedResource Include="Plugin/**/*">
```
[the remark this answers is not in the document]
Comment (markwallace-microsoft) [reply dotnet/src/Extensions/Planning.StepwisePlanner/Planning.StepwisePlanner.csproj:25]: Should be `Plugin`, the folder represents a single semantic plugin

**Review by SergeyMenshykh — APPROVED** [review]

**Review by SergeyMenshykh — APPROVED** [review]

**Pull request merged**